pkg/logentry/stages/multiline.go

```diff
@@ -0,0 +1,210 @@
+
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/promtail/api"
+)
+
+const (
+	ErrMultilineStageEmptyConfig        = "multiline stage config must define `firstline` regular expression"
+	ErrMultilineStageInvalidRegex       = "multiline stage first line regex compilation error: %v"
+	ErrMultilineStageInvalidMaxWaitTime = "multiline stage `max_wait_time` parse error: %v"
+)
+
+const (
+	maxLineDefault uint64 = 128
+	maxWaitDefault = 3 * time.Second
+)
+
+// MultilineConfig contains the configuration for a multilineStage
+type MultilineConfig struct {
+	Expression  *string `mapstructure:"firstline"`
+	regex       *regexp.Regexp
+	MaxLines    *uint64 `mapstructure:"max_lines"`
+	MaxWaitTime *string `mapstructure:"max_wait_time"`
+	maxWait     time.Duration
+}
+
+func validateMultilineConfig(cfg *MultilineConfig) error {
+	if cfg == nil || cfg.Expression == nil {
+		return errors.New(ErrMultilineStageEmptyConfig)
+	}
+
+	expr, err := regexp.Compile(*cfg.Expression)
+	if err != nil {
+		return errors.Errorf(ErrMultilineStageInvalidRegex, err)
+	}
+	cfg.regex = expr
+
+	if cfg.MaxWaitTime != nil {
+		maxWait, err := time.ParseDuration(*cfg.MaxWaitTime)
+		if err != nil {
+			return errors.Errorf(ErrMultilineStageInvalidMaxWaitTime, err)
+		}
+		cfg.maxWait = maxWait
+	} else {
+		cfg.maxWait = maxWaitDefault
+	}
+
+	if cfg.MaxLines == nil {
+		cfg.MaxLines = new(uint64)
+		*cfg.MaxLines = maxLineDefault
+	}
+
+	return nil
+}
+
+// dropMultiline matches lines to determine whether the following lines belong to a block and should be collapsed
+type multilineStage struct {
+	logger log.Logger
+	cfg    *MultilineConfig
+}
+
+// multilineState captures the internal state of a running multiline stage.
+type multilineState struct {
+	buffer         *bytes.Buffer // The lines of the current multiline block.
+	startLineEntry Entry         // The entry of the start line of a multiline block.
+	currentLines   uint64        // The number of lines of the current multiline block.
+}
+
+// newMulitlineStage creates a MulitlineStage from config
+func newMultilineStage(logger log.Logger, config interface{}) (Stage, error) {
+	cfg := &MultilineConfig{}
+	err := mapstructure.WeakDecode(config, cfg)
+	if err != nil {
+		return nil, err
+	}
+	err = validateMultilineConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &multilineStage{
+		logger: log.With(logger, "component", "stage", "type", "multiline"),
+		cfg:    cfg,
+	}, nil
+}
+
+func (m *multilineStage) Run(in chan Entry) chan Entry {
+	out := make(chan Entry)
+	go func() {
+		defer close(out)
+
+		streams := make(map[model.Fingerprint](chan Entry))
+		wg := new(sync.WaitGroup)
+
+		for e := range in {
+			key := e.Labels.FastFingerprint()
+			s, ok := streams[key]
+			if !ok {
+				// Pass through entries until we hit first start line.
+				if !m.cfg.regex.MatchString(e.Line) {
+					level.Debug(m.logger).Log("msg", "pass through entry", "stream", key)
```

Comment:
This sucks but all debug logs that are in the hot path should be using:

```
if Debug{
}
```

This is because the logger is really not performant, even if the debug log is deactivated. (expensation of strings still happen).

see https://github.com/grafana/loki/blob/master/cmd/promtail/main.go#L85